### PR TITLE
Implement missing constructors outlined by LWG issue 2713

### PR DESCRIPTION
**Pull request by cmazakas**
Closes https://github.com/boostorg/unordered/issues/154

### CODE REVIEW

File: include/boost/unordered/unordered_map.hpp [hunk 1]
```diff
@@ -1751,6 +1762,16 @@ namespace boost {
     {
     }
 
+    template <class K, class T, class H, class P, class A>
+    template <class InputIterator>
+    unordered_map<K, T, H, P, A>::unordered_map(
+      InputIterator f, InputIterator l, const allocator_type& a)
+        : table_(boost::unordered::detail::initial_size(f, l, 0), hasher(),
```
Comment (pdimov) [inline include/boost/unordered/unordered_map.hpp:1769]: This passes 0 for the initial bucket count...

File: include/boost/unordered/unordered_map.hpp [hunk 2]
```diff
@@ -1773,6 +1794,16 @@ namespace boost {
 
 #if !defined(BOOST_NO_CXX11_HDR_INITIALIZER_LIST)
 
+    template <class K, class T, class H, class P, class A>
+    unordered_map<K, T, H, P, A>::unordered_map(
+      std::initializer_list<value_type> list, const allocator_type& a)
+        : table_(boost::unordered::detail::initial_size(
+                   list.begin(), list.end(), detail::default_bucket_count),
```
Comment (pdimov) [inline include/boost/unordered/unordered_map.hpp:1801]: ... but this passes detail::default_bucket_count. Is this intended?

Comment (pdimov) [reply include/boost/unordered/unordered_map.hpp:1801]: The other two places both use detail::default_bucket_count, so probably not.

File: test/unordered/constructor_tests.cpp [hunk 3]
```diff
@@ -317,66 +329,208 @@ namespace constructor_tests {
...
-    std::initializer_list<typename T::value_type> list;
+    typedef typename T::value_type value_type;
+
+    std::initializer_list<value_type> list;
+
+    test::random_values<T> v(3, generator);
+    std::vector<value_type> vec(v.begin(), v.end());
+    BOOST_ASSERT(vec.size() >= 3);
+
+    std::initializer_list<value_type> nonempty_list{vec[0], vec[1], vec[2]};
```
Comment (pdimov) [inline test/unordered/constructor_tests.cpp:340]: This looks fishy to me. The backing array of the initializer_list is destroyed at the end of the full expression.

File: test/unordered/constructor_tests.cpp [hunk 4]
```diff
@@ -317,66 +329,208 @@ namespace constructor_tests {
...
+        T x(list);
+        BOOST_TEST(x.empty());
+        BOOST_TEST_EQ(x.bucket_count(), 0u);
+        BOOST_TEST(test::equivalent(x.hash_function(), hf));
+        BOOST_TEST(test::equivalent(x.key_eq(), eq));
+        BOOST_TEST(test::equivalent(x.get_allocator(), al));
+      }
+
+      {
+        T x(nonempty_list);
```
Comment (pdimov) [inline test/unordered/constructor_tests.cpp:361]: This should probably be `T x{ vec[0], vec[1], vec[2] };` instead.

File: test/unordered/constructor_tests.cpp (hunk 3 above)
Comment (cmazakas) [reply test/unordered/constructor_tests.cpp:340]: Hmm, that's a good question.

It seems to pass both sanitizers and checks for the values. For example, we still pass the check where we compare the container's contents to `expected`.

File: test/unordered/constructor_tests.cpp (hunk 4 above)
Comment (cmazakas) [reply test/unordered/constructor_tests.cpp:361]: This is a nice and simple refactor; will do!

File: include/boost/unordered/unordered_map.hpp (hunk 2 above)
Comment (cmazakas) [reply include/boost/unordered/unordered_map.hpp:1801]: Good catch, we should be using `default_bucket_count` everywhere.

File: test/unordered/constructor_tests.cpp (hunk 3 above)
Comment (pdimov) [reply test/unordered/constructor_tests.cpp:340]: Yeah, declaring a variable of type initializer_list does extend the lifetime of the temporary backing array, https://eel.is/c++draft/dcl.init.list#6. So it's fine.

**Review by pdimov — APPROVED** [review]

**Comment by pdimov** [comment]
This is going into 1.81, right?

**Comment by joaquintides** [comment]
> This is going into 1.81, right?

Let's decide, because if it goes in 1.81 I have to update the docs. I'm for it.

**Comment by pdimov** [comment]
We still have a few days, but you need to review and merge this to develop before I can merge it to master.

**Review by joaquintides — APPROVED** [review]
Hi Chris, I think the `boost::unordered::detail::default_bucket_count ` construct should be mimicked in foa with a dedicated, separate `boost::unordered::detail::foa::default_bucket_count`. As this is not directly related to the PR at hand, I'm approving but please keep this in your backlog. Thank you!

**Pull request merged**